### PR TITLE
JVM-1866: merge allocation state for the dead local variables.

### PR DESCRIPTION
**Pull request by navyxliu**
It seems to be counter-intuitive to say that we need to merge allocation states for the dead variables.There are 2 reasons. 1) method arguments and 2) indirect load.

1. method argument

in JVMSpec, 2.6.1. Local Variables, [reference](https://docs.oracle.com/javase/specs/jvms/se20/html/jvms-2.html#jvms-2.6.1)
>  The Java Virtual Machine uses local variables to pass parameters on method invocation. On class method invocation, any parameters are passed in consecutive local variables starting from local variable 0. On instance method invocation, local variable 0 is always used to pass a reference to the object on which the instance method is being invoked (this in the Java programming language). Any parameters are subsequently passed in consecutive local variables starting from local variable 1.

In the following example, the constructor method MergeAllVirts() has 2 arguments:(MergeAllVirts this, boolean cond).
In Parse, they are treated as normal local variables. Therefore, LV0 and LV1 have already been dead at the SafePointNode blackhole(). 

However, we need to consider them because the lifecycle of this is not over yet. caller keeps using this even after this ctor returns.

```java
class MergeAllVirts {
    int a;
    int b;
    int c = 100;

    public MergeAllVirts(boolean cond) {
        if (cond) {
            this.a = 1;
            this.b = 20;
            this.c = 300;
            blackhole(); // kill locals, so LV0 = this is dead here.
        } else {
            this.a = 100;
            this.b = 50;
            this.c = 7;
            blackhole(); // kill locals.
        }
        // merge 2 predecessors, but LV0 is not live here.
        // we need to merge allocation states, or wrong current allocation state is wrong.
    }
...
}
```
2. indirect load

The second case is even trickier. Currently, we don't intercept get-field bytecode. (c2 ideal graph misses the equivalent load-field node in Graal HIR).

It's possible that we will use an object later by an indirect load. therefore, we need to merge all objects no matter it is live or dead.

### CODE REVIEW

File: src/hotspot/share/opto/partialEscape.hpp [hunk 1]
```diff
@@ -106,8 +104,7 @@ class EscapedState: public ObjectState {
   ObjectState* clone() const override {
     return new EscapedState(_materialized);
   }
-  ObjectState& merge(ObjectState& newin,GraphKit* kit, const TypeOopPtr* oop_type,
-                     RegionNode* region, int pnum) override {
+  ObjectState& merge(ObjectState* newin,GraphKit* kit, RegionNode* region, int pnum) override {
```
Comment (caojoshua) [inline src/hotspot/share/opto/partialEscape.hpp]: Missing space before `GraphKit*`

File: src/hotspot/share/opto/partialEscape.hpp [hunk 2]
```diff
@@ -176,4 +189,15 @@ class PEAState {
 #endif
 };
 
+class MergeProcessor {
```
Comment (caojoshua) [inline src/hotspot/share/opto/partialEscape.hpp]: Should we call this `PEAMergeProcessor`? MergeProcessor feels a too generic

File: src/hotspot/share/opto/parseHelper.cpp [hunk 3]
```diff
@@ -680,3 +679,39 @@ void PEAState::validate() const {
...
+      vs1->merge(vs2, kit, region, pnum);
+    }
+  }
+}
+
+MergeProcessor::~MergeProcessor() {
+
+}
+
+void MergeProcessor::process_phi(PhiNode* phi) {
```
Comment (caojoshua) [inline src/hotspot/share/opto/parseHelper.cpp]: what is this function for?

File: src/hotspot/share/opto/parseHelper.cpp [hunk 4]
```diff
@@ -680,3 +679,39 @@ void PEAState::validate() const {
...
+  // intersection of two aliases
+  Unique_Node_List alias1;
+  Unique_Node_List alias2;
+
+  _state.aliases()->iterate([&](Node* alias, ObjID obj) {
+                             alias1.push(alias);
+                             return true;
+                           });
+  newin.aliases()->iterate([&](Node* alias, ObjID obj) {
+                            alias2.push(alias);
```
Comment (caojoshua) [inline src/hotspot/share/opto/parseHelper.cpp]: If we are going to `and` the two alias lists together, can't we just push to `alias1` here?

Comment (navyxliu) [reply src/hotspot/share/opto/parseHelper.cpp]: I intend to have the intersection of 2 aliases set. 
alias1 = _state.aliases()->keys();
alias2 = newin.aliases()->key();

```
  VectorSet set = alias1.member_set();
  set &= alias2.member_set();
  alias1.remove_useless_nodes(set);
```

File: src/hotspot/share/opto/parseHelper.cpp (hunk 3 above)
Comment (navyxliu) [reply src/hotspot/share/opto/parseHelper.cpp]: This is a placeholder function. I plan to use it in the loopbody of merge_common. 
yes, I better clean up everything before merging it.

File: src/hotspot/share/opto/partialEscape.hpp (hunk 2 above)
Comment (navyxliu) [reply src/hotspot/share/opto/partialEscape.hpp]: reasonable.

**Review by caojoshua — APPROVED** [review]

**Pull request merged**